### ISSUE TITLE
Server side rendering fails with identifier 'Set' undefined

### ISSUE THREAD

**Issue by HorizonShadow** [reactjs/react-rails#947]
### Steps to reproduce

* Make any old component
* Set prerender to true
* Try to render the component

### Expected behavior

It should pre-render the component.

### Actual behavior

Execjs crashes with the following error:

```  
ExecJS::ProgramError (identifier 'Set' undefined):
  
(execjs):1
app/controllers/entry_point_controller.rb:3:in `index'
```

### System configuration
Sprockets or Webpacker version: webpacker (3.5.5)
React-Rails version: react-rails (2.4.7)
React_UJS version: "react_ujs": "^2.4.4"
Rails version: rails (5.2.1)
Ruby version: RUBY VERSION: 2.4.4 (2018-03-28 patchlevel 296) [i386-mingw32]
-------

Basically everything works perfectly fine, but if I attempt to use server-side rendering, the program crashes with the above error: `identifier 'Set' undefined`. I thought it might have been due to an outdated, but I updated to the latest node and the issue persists.

**Comment by BookOfGreg** [comment]
Marked invalid as your provided reproduction does not cause the effect you say for me.

Could you please provide code that exhibits the issue you are seeing?

**Comment by HorizonShadow** [comment]
Yes, with prerender set to false, everything runs correctly. 

I'll create a new project and detail everything I do.

* `rails new react-rails-prerender-test --webpack=react --skip --database=postgresql`
* `bundle add react-rails`
* `rails g react:install`
* `rails g controller entry index -s`
* `rails g react:component Entry -s`
* Changed the render method of Entry.js to say "Hello World!"
* Added `    render_component :Entry` to the index route of EntryController
* `rails server -b 127.0.0.1 -p 3000 -e development`
* Navigate to localhost:3000/entry/index
```
Started GET "/entry/index" for 127.0.0.1 at 2018-10-19 07:02:03 -0300
Processing by EntryController#index as HTML
[Webpacker] Compilingâ€¦
[Webpacker] Compiled all packs in C:/Users/horiz/RubymineProjects/react-rails-prerender-test/public/packs
Completed 500 Internal Server Error in 3005ms (ActiveRecord: 0.0ms)


  
ExecJS::ProgramError (identifier 'Set' undefined):
  
(execjs):1
app/controllers/entry_controller.rb:3:in `index'
```
![image](https://user-images.githubusercontent.com/1729810/47211900-21fe2580-d36d-11e8-8ac9-de0cba32d986.png)

**Comment by BookOfGreg** [comment]
I followed your instructions but I still do not have your issue, here is my bash history:
```
509  rails new foo --webpack=react
  510  cd foo
  511  bundle add react-rails
  512  rails g react_install
  513  rails g react:install
  514  rails g controller -h
  515  rails g controller bar index
  516  rails g react:component Entry
  517  code . # added render component: :Entry and <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
  518  rails s
```

Still cannot reproduce your issue.

**Comment by BookOfGreg** [comment]
OH! You're using your system NodeJS
Do `bundle add miniracer` and it will start working for you.
`Execjs` automatically loads system JS runtime, which is Chakra on Windows, JSCore on OSX, and usually NodeJS if available. We only support Node > 6, therefore we do not support `therubyracer` as it's using libv8-v3 which is far too old.

Upgrade your node to > 6, or include `miniracer`.

**Comment by BookOfGreg** [comment]
If you are seeing the 'Set undefined' issue on a modern runtime that would be most surprising to me.
What do you see when checking the execjs runtime?
```
[2] pry(main)> ExecJS.runtime
=> #<ExecJS::ExternalRuntime:0x00007f834db1d730
 @binary="node",
 @command=["nodejs", "node"],
 @deprecated=false,
 @encoding="UTF-8",
 @name="Node.js (V8)",
 @popen_options={:external_encoding=>"UTF-8", :internal_encoding=>#<Encoding:UTF-8>},
 @runner_path=
  "/Users/my_user/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/execjs-2.7.0/lib/execjs/support/node_runner.js">
```

**Comment by HorizonShadow** [comment]
It's trying to use the "Duktape" runtime.

I'm pretty sure miniracer isn't compatible with windows, given that it checks for node with `which node`, and that command doesn't exist on windows.

I'm trying to get it running on WSL now, but it's still using the Duktape runtime.

**Comment by BookOfGreg** [comment]
Wow, Duktape. That's a rare runtime, unfortunately it only supports ES5 by the look of it's readme so it's not going to be one we support.

Still if you can't use miniracer due to it not working on Windows that's a problem, that means that Node would be the only Windows compatible JS runtime and miniracer is about 10x faster than NodeJS.
Out of curiosity why do you program on Windows?

Now that you've got to the next issue, it looks like it's still a JS related issue:
https://stackoverflow.com/questions/34130539/uncaught-error-invariant-violation-element-type-is-invalid-expected-a-string
This doesn't match your perfectly but is close, and there are lots of results on google for similar issues, mostly Webpack related.

**Comment by HorizonShadow** [comment]
I think it's a bit broader than just a JS issue, since it works when not pre-rendering. Probably a dependency problem that can't be pre-rendered. Haven't had a chance to look at it again yet.

Regarding programming on windows, I simply don't have a linux box and can't be assed to dual boot. Ruby programming on windows has gotten a lot better in recent years, but clearly it's not there yet.

**Comment by rahulakurati** [comment]
Even, I could able to reproduce this
> ReferenceError: Set is not defined
>   (execjs):35:458
when I'm using server-side rendering (including `{prerender: true}` in the helper method). However, client side rendering is working good. Any suggestions for fixing it?

**Comment by BookOfGreg** [comment]
@rahulakurati Glad that solved it for you.
